File: src/main/java/io/trydent/olimpo/sys/Property.java
package io.trydent.olimpo.sys;

import io.trydent.olimpo.type.Type;

public interface Property extends Type.AsString {
  static Property envVar(final String name) {
    return new EnvVar(name);
  }
}

final class EnvVar implements Property {
  private final String name;

  EnvVar(String name) {
    this.name = name;
  }

  @Override
  public final String get() {
    return System.getenv(name);
  }
}

File: src/main/java/io/trydent/olimpo/sys/Property.kt
package io.trydent.olimpo.sys

interface Property : () -> String? {
  companion object {
    fun envVar(name: String): Property = EnvVar(name)

    fun system(property: String): Property = SystemProperty(property)
  }
}

internal class EnvVar(private val name: String) : Property {
  override fun invoke(): String? = System.getenv(name)
}

internal class SystemProperty(private val name: String) : Property {
  override fun invoke(): String? = System.getProperty(name)
}
